runtime/lua/vim/lsp/buf.lua

```diff
@@ -811,4 +811,19 @@ function M.inlay_hint(bufnr, enable)
   end
 end
 
+--- Enable/disable/toggle diagnostics for a buffer
+---@param bufnr (integer) Buffer handle, or 0 for current
+---@param enable (boolean|nil) true/false to enable/disable, nil to toggle
+function M.diagnostic(bufnr, enable)
```

Comment:
Given that `vim.lsp.diagnostic` is already public I don't think we'll want to have this here in `buf` with 3 new additional public functions in `vim.lsp.diagnostic` (enable, disable, toggle).

I think diagnostic should be enabled by default given that current push diagnostics already are.
That means we may not need a new public entry point at all, but let users do the (display) toggle via `vim.diagnostic`

(Downside is that users won't be able to prevent the requests from happening if they really do not want diagnostic, but diagnostic are required for other functionality like code-actions, so I think they'll just have to pay the costs - if the server isn't pushing diagnostics anyways)

Reply:
In this case, would it make sense to set up the the textDocument/diagnostic handlers within `vim.lsp.buf_attach_client` ?